### PR TITLE
Automated the manual node registration.

### PR DESCRIPTION
**Pull request by PetiAPocok**
Automated the manual node registration and handled the nodes with nodebox and mesh drawtype. (They are excluded.)

### CODE REVIEW

File: init.lua [hunk 1]
```diff
@@ -209,293 +209,16 @@ minetest.register_craft({
...
-	"binoculars:binoculars",
-	"map:mapping_kit",
-}) do
-	frame.register(node)
+local items_to_frame = {}
+
+for i, node in pairs(minetest.registered_items) do
+    if minetest.get_item_group(node.name, "not_in_creative_inventory") == 0 and node.drawtype ~= "airlike" and node.drawtype ~= "nodebox" and node.drawtype ~= "mesh" then
+        items_to_frame[i] = node.name
+    elseif (node.drawtype == "nodebox" or node.drawtype == "mesh") and node.inventory_image ~= "" then
```
Comment (Panquesito7) [inline init.lua]:
```suggestion
    elseif (node.drawtype == "nodebox"
        or node.drawtype == "mesh") and node.inventory_image ~= "" then
```

File: init.lua [hunk 2]
```diff
@@ -209,293 +209,16 @@ minetest.register_craft({
...
-	"xpanes:pane_flat",
-	"xpanes:obsidian_pane",
-	"binoculars:binoculars",
-	"map:mapping_kit",
-}) do
-	frame.register(node)
+local items_to_frame = {}
+
+for i, node in pairs(minetest.registered_items) do
+    if minetest.get_item_group(node.name, "not_in_creative_inventory") == 0 and node.drawtype ~= "airlike" and node.drawtype ~= "nodebox" and node.drawtype ~= "mesh" then
```
Comment (Panquesito7) [inline init.lua]:
```suggestion
    if minetest.get_item_group(node.name, "not_in_creative_inventory") == 0
        and node.drawtype ~= "airlike" and node.drawtype
        ~= "nodebox" and node.drawtype ~= "mesh" then
```

**Review by Panquesito7 — APPROVED** [review]
:+1: apart from the above.

**Comment by Panquesito7** [comment]
@SmallJoker this PR looks fine, it can be merged.

**Comment by SmallJoker** [comment]
@sofar ?

**Comment by HeyITGuyFixIt** [comment]
Tried it out, and a game failed to load with this error:
```
ModError: Failed to load and run script from /home/csirolli/.minetest/mods/frame/init.lua:
/home/csirolli/.minetest/mods/frame/init.lua:139: attempt to index field 'tiles' (a nil value)
stack traceback:
	/home/csirolli/.minetest/mods/frame/init.lua:139: in function 'register'
	/home/csirolli/.minetest/mods/frame/init.lua:227: in main chunk
Check debug.txt for details.
```

This is in the final else part.

**Comment by sofar** [comment]
Well, so, thinking about this, I'm not sure about it.

If we make this dynamic, due to the random nature of dependency ordering in minetest, it's possible that mods randomly will be "spotted" by this new code and the next time it runs not. That will cause unknown item errors. We know the ordering can change when something as simple as another mod being added that isn't consequential for the ordering of this mod, it just changes. The only guarantee in ordering is that optional_depends mods are loaded before frame. But there is no guarantee (obviously) that other mods are _not_ loaded before frame.

If this code explicitly only handles mods from minetest_game, then that would not be a problem. But it doesn't.

The error reported by @HeyITGuyFixIt may actually be such a failure/case.

So, I'm leaning towards "this isn't acceptable - it will randomly break stuff for users that is difficult to explain".

**Comment by HeyITGuyFixIt** [comment]
I'd have to experiment some to figure out which mod caused the issue.

**Comment by sofar** [comment]
It's entirely possible that there's a bug in the patch, too, btw. I haven't tested this.

File: init.lua [hunk 3]
```diff
@@ -209,293 +209,19 @@ minetest.register_craft({
...
+local items_to_frame = {}
+
+for i, node in pairs(minetest.registered_items) do
+    if minetest.get_item_group(node.name, "not_in_creative_inventory") == 0
+        and node.drawtype ~= "airlike" and node.drawtype
+        ~= "nodebox" and node.drawtype ~= "mesh" then
+        items_to_frame[i] = node.name
+    elseif (node.drawtype == "nodebox"
+        or node.drawtype == "mesh") and node.inventory_image ~= "" then
+        items_to_frame[i] = node.name
```
Comment (louisroyer) [inline init.lua]:
```suggestion
        table.insert(items_to_frame, node.name)
```

File: init.lua [hunk 4]
```diff
@@ -209,293 +209,19 @@ minetest.register_craft({
...
-}) do
-	frame.register(node)
+local items_to_frame = {}
+
+for i, node in pairs(minetest.registered_items) do
+    if minetest.get_item_group(node.name, "not_in_creative_inventory") == 0
+        and node.drawtype ~= "airlike" and node.drawtype
+        ~= "nodebox" and node.drawtype ~= "mesh" then
+        items_to_frame[i] = node.name
+    elseif (node.drawtype == "nodebox"
```
Comment (louisroyer) [inline init.lua]:
```suggestion
        or ((node.drawtype == "nodebox"
```

File: init.lua [hunk 5]
```diff
@@ -209,293 +209,19 @@ minetest.register_craft({
...
-	"map:mapping_kit",
-}) do
-	frame.register(node)
+local items_to_frame = {}
+
+for i, node in pairs(minetest.registered_items) do
+    if minetest.get_item_group(node.name, "not_in_creative_inventory") == 0
+        and node.drawtype ~= "airlike" and node.drawtype
+        ~= "nodebox" and node.drawtype ~= "mesh" then
+        items_to_frame[i] = node.name
```
Comment (louisroyer) [inline init.lua]:
```suggestion
```

File: init.lua [hunk 6]
```diff
@@ -209,293 +209,19 @@ minetest.register_craft({
...
-	"xpanes:pane_flat",
-	"xpanes:obsidian_pane",
-	"binoculars:binoculars",
-	"map:mapping_kit",
-}) do
-	frame.register(node)
+local items_to_frame = {}
+
+for i, node in pairs(minetest.registered_items) do
+    if minetest.get_item_group(node.name, "not_in_creative_inventory") == 0
```
Comment (louisroyer) [inline init.lua]:
```suggestion
    if (minetest.get_item_group(node.name, "not_in_creative_inventory") == 0
```

File: init.lua [hunk 7]
```diff
@@ -209,293 +209,19 @@ minetest.register_craft({
...
+        and node.drawtype ~= "airlike" and node.drawtype
+        ~= "nodebox" and node.drawtype ~= "mesh" then
```
Comment (louisroyer) [inline init.lua]:
```suggestion
        ~= "nodebox" and node.drawtype ~= "mesh")
```

File: init.lua [hunk 8]
```diff
@@ -209,293 +209,19 @@ minetest.register_craft({
...
+        items_to_frame[i] = node.name
+    elseif (node.drawtype == "nodebox"
+        or node.drawtype == "mesh") and node.inventory_image ~= "" then
```
Comment (louisroyer) [inline init.lua]:
```suggestion
        or node.drawtype == "mesh") and node.inventory_image ~= "") then
```

File: init.lua [hunk 9]
```diff
@@ -209,293 +209,19 @@ minetest.register_craft({
...
-	"xpanes:bar_flat",
-	"xpanes:pane_flat",
-	"xpanes:obsidian_pane",
-	"binoculars:binoculars",
-	"map:mapping_kit",
-}) do
-	frame.register(node)
+local items_to_frame = {}
+
+for i, node in pairs(minetest.registered_items) do
```
Comment (louisroyer) [inline init.lua]:
```suggestion
for _, node in pairs(minetest.registered_items) do
```

**Review by louisroyer — COMMENTED** [review]
Maybe you could also add a filter based on the prefix of `node.name` (to register only if the node is part of mtg or is `frame:empty`).

File: init.lua (hunk 6 above)
Comment (Panquesito7) [reply init.lua]: Missing `)` at the end, otherwise it'll throw error.

Comment (louisroyer) [reply init.lua]: is the one I suggested line 217 not enough? (I don't know how to make a multi-lines suggestion and this is not very convenient to read)

**Comment by HeyITGuyFixIt** [comment]
The error I am getting is coming from the [lava_ore_gen](https://content.minetest.net/packages/Coder12/lava_ore_gen/) mod. It redefines `default:stone` and creates it's own `lava_ore_gen:stone_hot`. The `lava_ore_gen:stone_hot` has its own tile, which the error suggests is a nil value. I wouldn't expect the tile for `default:stone` to be removed when `minetest.override_item` is called, especially when tiles aren't modified by it (only `floodable` and `on_flood` are modified). So idk what the issue is here.

For reference, this is some of the relevant code:

```lua
local function create_hotstone()
	minetest.register_node("lava_ore_gen:stone_hot", {name = "lava_ore_gen:stone_hot"})
end
local function override_hotstone()
	local clonenodeForLava = {}
	-- [...]
	clonenodeForLava.tiles = {"default_stone.png^[colorize:red:20"}
	-- [...]
	minetest.register_node(":lava_ore_gen:stone_hot", clonenodeForLava)
end
-- [...]
create_hotstone()
minetest.register_on_mods_loaded(function()
	override_hotstone()
	override_stone()
end)
```

I omitted the `override_stone` function since, as mentioned earlier, it only modifies the `floodable` and `on_flood` properties of `default:stone`.

File: init.lua (hunk 6 above)
Comment (HeyITGuyFixIt) [reply init.lua]: The one on 217 is enough. I copied your changes into my local copy and I don't get any errors from this.

**Comment by HeyITGuyFixIt** [comment]
I think I see the issue with that mod. They register the node, but without a tile, and then add onto it when all the mods have been loaded. So when frames sees it, it doesn't have a tile.

File: init.lua (hunk 3 above)
Comment (PetiAPocok) [reply init.lua]: One of the automation's goal is to automatically add new items and blocks. So I'm against the prefix filtering.

**Comment by PetiAPocok** [comment]
Well we could check if the node has tiles... But what is a tileless node? Or if it doesn't have sides then why isn't it airlike?

**Comment by HeyITGuyFixIt** [comment]
> Well we could check if the node has tiles... But what is a tileless node? Or if it doesn't have sides then why isn't it airlike?

In this case, the mod author registered the node without a tile, and then overrode it's properties later to include a tile. Not sure why, but that's what they did. 

Idk if you can listen for when a node is overridden and then update its appropriate frame.

**Comment by MoNTE48** [comment]
https://github.com/minetest-mods/flowerpot/pull/4#issuecomment-563739041

**Comment by sofar** [comment]
I have not seen anything convincing me that this is fully safe even though I explained why it can't be. I don't think this should be merged until that problem is solved.

**Comment by PetiAPocok** [comment]
And what defines the mod's load order, because it is clearly not alphabetical... The dependencies?

`2020-07-08 16:40:31: INFO[Main]:   [game_commands] ["/home/user/.minetest/games/minetest_extended/mods/game_commands/init.lua"]

2020-07-08 16:40:31: INFO[Main]: Mod "game_commands" loaded after 0 seconds
2020-07-08 16:40:31: INFO[Main]:   [effects_hud ] ["/home/user/.minetest/games/minetest_extended/mods/effects_hud/init.lua"]
2020-07-08 16:40:31: INFO[Main]: Mod "effects_hud" loaded after 0 seconds
2020-07-08 16:40:31: INFO[Main]:   [alchemy     ] ["/home/user/.minetest/games/minetest_extended/mods/alchemy/init.lua"]
2020-07-08 16:40:31: INFO[Main]: Mod "alchemy" loaded after 0 seconds
2020-07-08 16:40:31: INFO[Main]:   [dye         ] ["/home/user/.minetest/games/minetest_extended/mods/dye/init.lua"]
2020-07-08 16:40:31: INFO[Main]: Mod "dye" loaded after 0.001 seconds
2020-07-08 16:40:31: INFO[Main]:   [wool        ] ["/home/user/.minetest/games/minetest_extended/mods/wool/init.lua"]
2020-07-08 16:40:31: INFO[Main]: Mod "wool" loaded after 0 seconds
2020-07-08 16:40:31: INFO[Main]:   [paintings   ] ["/home/user/.minetest/games/minetest_extended/mods/paintings/init.lua"]
2020-07-08 16:40:31: INFO[Main]: Mod "paintings" loaded after 0.001 seconds
2020-07-08 16:40:31: INFO[Main]:   [farming     ] ["/home/user/.minetest/games/minetest_extended/mods/farming/init.lua"]
2020-07-08 16:40:31: INFO[Main]: Mod "farming" loaded after 0.051 seconds
2020-07-08 16:40:31: INFO[Main]:   [mobs        ] ["/home/user/.minetest/games/minetest_extended/mods/mobs/init.lua"]
2020-07-08 16:40:31: ACTION[Main]: [MOD] Mobs Redo loaded
2020-07-08 16:40:31: INFO[Main]: Mod "mobs" loaded after 0.009 seconds
2020-07-08 16:40:31: INFO[Main]:   [mobs_monster] ["/home/user/.minetest/games/minetest_extended/mods/mobs_monster/init.lua"]
2020-07-08 16:40:31: INFO[Main]: Mod "mobs_monster" loaded after 0.011 seconds
2020-07-08 16:40:31: INFO[Main]:   [mobs_animal ] ["/home/user/.minetest/games/minetest_extended/mods/mobs_animal/init.lua"]
2020-07-08 16:40:31: INFO[Main]: Mod "mobs_animal" loaded after 0.017 seconds
2020-07-08 16:40:31: INFO[Main]:   [backpacks   ] ["/home/user/.minetest/games/minetest_extended/mods/backpacks/init.lua"]
2020-07-08 16:40:31: INFO[Main]: Mod "backpacks" loaded after 0 seconds
2020-07-08 16:40:31: INFO[Main]:   [flowerpot   ] ["/home/user/.minetest/games/minetest_extended/mods/flowerpot/init.lua"]
2020-07-08 16:40:31: INFO[Main]: Mod "flowerpot" loaded after 0.001 seconds
2020-07-08 16:40:31: INFO[Main]:   [fishing     ] ["/home/user/.minetest/games/minetest_extended/mods/fishing/init.lua"]
2020-07-08 16:40:31: INFO[Main]: Mod "fishing" loaded after 0.004 seconds
2020-07-08 16:40:31: INFO[Main]:   [hbhunger    ] ["/home/user/.minetest/games/minetest_extended/mods/hbhunger/init.lua"]
2020-07-08 16:40:31: INFO[Main]: hbhunger: Previously saved data loaded.
2020-07-08 16:40:31: ACTION[Main]: hb.register_hudbar: satiation
2020-07-08 16:40:31: INFO[Main]: Mod "hbhunger" loaded after 0.001 seconds
2020-07-08 16:40:31: INFO[Main]:   [hbsprint    ] ["/home/user/.minetest/games/minetest_extended/mods/hbsprint/init.lua"]
2020-07-08 16:40:31: INFO[Main]: Mod "hbsprint" loaded after 0 seconds
2020-07-08 16:40:31: INFO[Main]:   [throwing_arrows] ["/home/user/.minetest/games/minetest_extended/mods/throwing/throwing_arrows/init.lua"]
2020-07-08 16:40:31: INFO[Main]: Mod "throwing_arrows" loaded after 0.001 seconds
2020-07-08 16:40:31: INFO[Main]:   [carpets     ] ["/home/user/.minetest/games/minetest_extended/mods/carpets/init.lua"]
2020-07-08 16:40:31: INFO[Main]: Mod "carpets" loaded after 0.002 seconds
2020-07-08 16:40:31: INFO[Main]:   [beds        ] ["/home/user/.minetest/games/minetest_extended/mods/beds/init.lua"]
2020-07-08 16:40:31: INFO[Main]: Mod "beds" loaded after 0.007 seconds
2020-07-08 16:40:31: INFO[Main]:   [spawn       ] ["/home/user/.minetest/games/minetest_extended/mods/spawn/init.lua"]
2020-07-08 16:40:31: INFO[Main]: Mod "spawn" loaded after 0 seconds
2020-07-08 16:40:31: INFO[Main]:   [map         ] ["/home/user/.minetest/games/minetest_extended/mods/map/init.lua"]
2020-07-08 16:40:31: INFO[Main]: Mod "map" loaded after 0 seconds
2020-07-08 16:40:31: INFO[Main]:   [frame       ] ["/home/user/.minetest/games/minetest_extended/mods/frame/init.lua"]
2020-07-08 16:40:31: INFO[Main]: Mod "frame" loaded after 0.054 seconds
2020-07-08 16:40:31: INFO[Main]:   [mesecons_lightstone] ["/home/user/.minetest/games/minetest_extended/mods/mesecons/mesecons_lightstone/init.lua"]
2020-07-08 16:40:31: INFO[Main]: Mod "mesecons_lightstone" loaded after 0.001 seconds
2020-07-08 16:40:31: INFO[Main]:   [clouds      ] ["/home/user/.minetest/games/minetest_extended/mods/clouds/init.lua"]
2020-07-08 16:40:31: INFO[Main]: Mod "clouds" loaded after 0 seconds
2020-07-08 16:40:31: INFO[Main]:   [worldedit   ] ["/home/user/.minetest/games/minetest_extended/mods/worldedit/worldedit/init.lua"]`

**Comment by sofar** [comment]
> And what defines the mod's load order, because it is clearly not alphabetical... The dependencies?

Yes, except that the ordering is not guaranteed because mods that are NOT a dependency may either be loaded BEFORE or AFTER the mod.

Suppose you have mod A depending on B and mod P depending on mod Q. The resulting sorting order can be any of the following:
- B A Q P
- B Q A P
- B Q P A
- Q B P A
- Q P B A

In other words, if A is `frame`, then from one run to another run it could see items from P or Q and make item frame nodes form them, and the next run it wouldn't, making them unknown nodes and likely causing a server crash or something worse.

There is no good solution to guard against this, which is why I'm against merging any "wildcard" node creation code like this, and why it wasn't added in the first place.

Honestly this PR should be closed before more time is wasted on trying to make it work.

**Comment by PetiAPocok** [comment]
A last question: what's with optional depend? If I were to put "all the supported mods" as optional dependency? Would it make the engine load them first, but if it isn't there then it won't be a problem. And check if a node has tiles before using its tiles... But in my opinion a tileless node should be airlike... But that is another topic...

**Comment by HeyITGuyFixIt** [comment]
@PetiAPocok you would need to be careful not to add anything that depends on frames, or depends on something that depends on frames. I'm not sure if it will be an issue with optionally depending on it and those mods optionally depending on us, but I think it will cause some issue. 

We would need to be careful to avoid circular dependencies.

**Comment by PetiAPocok** [comment]
Well, then this version will only be used my game, where I take care these problems.

**Comment by MoNTE48** [comment]
@PetiAPocok, I did the same with flowerpots.
Good developer =! Paranoid developer.

**Comment by sofar** [comment]
> Well, then this version will only be used my game, where I take care these problems.

That's acceptable - if you maintain your own `game` you should be able to decide this sort of thing for yourself. For mods that are generally used in many games, that's not doable.

**Pull request closed without merging**